File: 44.wildcard-matching.go
/*
 * @lc app=leetcode id=44 lang=golang
 *
 * [44] Wildcard Matching
 */

// @lc code=start
func isMatch(s string, p string) bool {
	// var dp [len(s) + 1][len(p) + 1]bool
	var dp [][]bool
	for i := 0; i < len(s)+1; i++ {
		var row []bool
		for j := 0; j < len(p)+1; j++ {
			row = append(row, false)
		}
		dp = append(dp, row)
	}
	dp[0][0] = true

	for i := 1; i <= len(p); i++ {
		if p[i-1] == '*' {
			dp[0][i] = dp[0][i-1]
		}
	}

	for i := 1; i < len(s)+1; i++ {
		for j := 1; j < len(p)+1; j++ {
			if p[j-1] == '?' {
				dp[i][j] = dp[i-1][j-1]
			} else if p[j-1] == '*' {
				dp[i][j] = dp[i-1][j] || dp[i][j-1]
			} else {
				dp[i][j] = dp[i-1][j-1] && (s[i-1] == p[j-1])
			}
		}
	}

	return dp[len(s)][len(p)]
}
// @lc code=end

File: 44.wildcard-matching.cpp
/*
 * @lc app=leetcode id=44 lang=cpp
 *
 * [44] Wildcard Matching
 */

// @lc code=start
class Solution {
 public:
  bool isMatch(string s, string p) {
    int m = s.size();
    int n = p.size();
    vector<vector<int>> memo(m + 1, vector<int>(n, -1));

    return dp(s, p, 0, 0, memo);
  }

 private:
  bool dp(const string& s, const string& p, int i, int j,
          vector<vector<int>>& memo) {
    if(j == p.size()) return i == s.size();

    if (memo[i][j] != -1) return static_cast<bool>(memo[i][j]);

    bool res = false;

    if (p[j] == '*') {
      res = (i < s.size() && dp(s, p, i + 1, j, memo)) || dp(s, p, i, j + 1, memo);
    } else {
      bool match = i < s.size() && (s[i] == p[j] || p[j] == '?');

      res = match && dp(s, p, i + 1, j + 1, memo);
    }


    return memo[i][j] = static_cast<int>(res);
  }
};
// @lc code=end

